### PR TITLE
Make \cdot and \cdotp operators equivalent

### PR DESCRIPTION
**Pull request by iagobaapellaniz**
It includes an enhancement discussed in #25098 

(See also https://discourse.julialang.org/t/inconsistencies-in-some-methods-and-more/7788)

### CODE REVIEW

**Comment by fredrikekre** [comment]
Shouldn't this be done as in https://github.com/JuliaLang/julia/pull/19464?

**Comment by i-apellaniz** [comment]
Probably. I'm not an expert on how to do this. 

I'll follow this thread and I'll study the PR #19464 in order to figure out how it can be done.

Thanks

**Comment by stevengj** [comment]
Yes, the right way to do this is to just add a line to https://github.com/JuliaLang/julia/blob/master/src/flisp/julia_charmap.h to make this substitution.

Also add a note to NEWS, and at the end of https://github.com/JuliaLang/julia/blob/master/doc/src/manual/variables.md#allowed-variable-names (along with the note about ε), and add a test.

**Comment by stevengj** [comment]
(Note also that you shouldn't open a new PR.   Just revert the commit in this branch and push new commits to fix it the right way.  It will all get squashed into a single commit when it is merged.  That way all of the discussion will go together.)

**Comment by i-apellaniz** [comment]
Yes. I'll do that :-) Thanks @stevengj

File: doc/src/manual/variables.md [hunk 1]
```diff
@@ -122,7 +122,8 @@ Different ways of entering Unicode combining characters (e.g., accents)
 are treated as equivalent (specifically, Julia identifiers are NFC-normalized).
 The Unicode characters `ɛ` (U+025B: Latin small letter open e)
 and `µ` (U+00B5: micro sign) are treated as equivalent to the corresponding
-Greek letters, because the former are easily accessible via some input methods.
+Greek letters, and '·' (U+00B7: Middle dot) is treated as the mathematical
```
Comment (stevengj) [inline doc/src/manual/variables.md]: Maybe link https://en.wikipedia.org/wiki/Interpunct

**Comment by stevengj** [comment]
Needs a NEWS item, then should be good to go.

**Comment by i-apellaniz** [comment]
Not yet, I think. I don't know if I'm missing something but Julia still complains about `\cdotp[TAB]` with
```jl
julia> ·
ERROR: syntax: invalid character "·"
```
If I edit the parser file `julia-parser.scm` to add a U+00b7 as in [59e73a8](https://github.com/JuliaLang/julia/pull/25157/commits/59e73a8c5d6a823a1cdacca90ec0a830f4f6ddb2), then it is no longer an invalid character but it is a non defined variable.
```jl
julia> ·
ERROR: UndefVarError: · not defined
```

**Comment by stevengj** [comment]
Oh, I know the problem.  The issue is that normalization of symbols only happens *after* they are parsed in the file, and since `·` is a punctuation character (Unicode category Po) it is not treated as a valid symbol.   This doesn't happen with `€` because `€` is already a valid identifier.

The fix is to add `·` to the list of operators at the top of `src/julia-parser.scm`.  Just add `'·` to the `prec-times` list.

**Comment by i-apellaniz** [comment]
I checked this and it does not work. Now it is a valid identifier but it seems that the mapping arrives before since
```jl
julia> ·
ERROR: UndefVarError: · not defined
```

Maybe this is necessary, but still does not fix it

**Comment by stevengj** [comment]
It could be that the normalization is not done to operators, let me check.

**Comment by stevengj** [comment]
Yup, the normalization is only done to variable names, not to operators.  There would need to be a call to a normalization routine at the end of `read-operator`.

**Comment by i-apellaniz** [comment]
Nice! With that I think I'm able to finish this. Thanks a lot!

File: src/julia-parser.scm [hunk 2]
```diff
@@ -549,7 +549,9 @@
                            op))
                         (else '|.|)))))
 
-          ((opchar? c)  (read-operator port (read-char port)))
+          ((if (opchar? c)
+              (accum-julia-symbol c port)
```
Comment (stevengj) [inline src/julia-parser.scm]: No, this is wrong; `accum-julia-symbol` reads an identifier (not an operator) from the input stream, leaving nothing else for `read-operator` to read.

I think you will need to export a new flisp function to normalize a symbol

Comment (i-apellaniz) [reply src/julia-parser.scm]: I added a new commit. It works now after `make cleanall` and `make`

**Comment by i-apellaniz** [comment]
It works now for me
```jl
julia> ⋅   # \cdotp[TAB] or U+00b7
dot (generic function with 9 methods)
julia> ·   # \cdot[TAB] or U+22c5
dot (generic function with 9 methods)
julia> dot
dot (generic function with 9 methods)
```

File: src/julia-parser.scm [hunk 3]
```diff
@@ -549,7 +549,9 @@
                            op))
                         (else '|.|)))))
 
-          ((opchar? c)  (read-operator port (read-char port)))
+          ((if (opchar? c)
+               (read-operator port (read-char port))
+               (accum-julia-symbol c port)))
```
Comment (JeffBezanson) [inline src/julia-parser.scm]: This prevents the following "invisible character" and "invalid character" errors from happening.

**Comment by stevengj** [comment]
Ping, what this the status here?

**Comment by i-apellaniz** [comment]
I'm not able to write anything that really works. I think we should work on the parser, which is written in SCM/C++. At the moment I have no time to learn all this.

It seems that there is a function in the parser that only works to make equivalent the identifiers, but not the operators. Help is needed, or a new PR should be opened to solve the issue.

**Comment by JeffBezanson** [comment]
I can help with this at some point, but it's not urgent since it's a non-breaking change.

**Comment by mlhetland** [comment]
Some relevant discussion [on Discourse](https://discourse.julialang.org/t/why-is-greek-ano-teleia-a-valid-identifier-character/12442).

**Comment by mlhetland** [comment]
A summary of the [Discourse discussion](https://discourse.julialang.org/t/why-is-greek-ano-teleia-a-valid-identifier-character/12442) based on my understanding:

- Greek ano teleia (U+0387) is a valid identifier character. This is motivated by it having the `Other_ID_Continue` property, as recommended by [UAX #31](http://unicode.org/reports/tr31/).
- Middle dot (U+00b7, `\cdotp`) is _not_ a valid identifier character, despite falling under the same UAX recommendation, and being [canonically equivalent](https://github.com/googlei18n/noto-fonts/issues/723#issuecomment-239917505) to the ano teleia.
- Julia currently normalizes the (valid) ano teleia into the (invalid) middle dot.
- Ano teleia, middle dot and dot operator look quite similar in many fonts; the first two are often identical (cf. `··⋅`).
- The middle dot is easy to typeon many keyboards.

**Comment by stevengj** [comment]
PR #28167 removes U+0387 from the list of valid identifiers.   If this PR is revived at some point, we will want to normalize both U+0387 and U+00b7 to U+22c5 (`\cdot`).

**Comment by stevengj** [comment]
I rebased this on `master` and fixed it by adding code to normalize operators as they are parsed, as I suggested above.

**Comment by i-apellaniz** [comment]
Thanks! This will help a lot of people using a spanish keyboard layout, since U+00B7 is just typed by pressing SHIFT+3. Thanks again

**Comment by i-apellaniz** [comment]
Could this make its way to 1.6?

**Comment by stevengj** [comment]
Too late for 1.6, I think.

**Comment by vtjnash** [comment]
Triage didn't seem to see why this was put on the list. It seemed like nobody had objected to merging?

**Pull request merged**